Feature/jenkins 35840 35781 more favoriting

cliffmeyers opened the pull request:
**Description**
* Allow favoriting of pipeline from the detail page (in the header), and ensure it reflects the correct favorite status.
* Allow favoriting of branches from the Branches tab on details page, and reflect current favorited status
* To test, you can toggle favoriting of items on these two screens and return to the dashboard to see which cards are displayed. Also deep-link directly to these screens to ensure the favorite stars are toggled on for the right pipelines or branches.
* I struggled with a way to do good unit tests for these. The bugs I found while developing were all integrated related, especially the deep-linking problems. I plan to cover these with acceptance tests in JENKINS-36641

**Submitter checklist**
- [x] Change is code complete and matches issue description
- [x] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [x] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

**Reviewer checklist**
- [x] Run the changes and verified the change matches the issue description
- [x] Reviewed the code
- [x] Verified that the appropriate tests have been written or valid explanation given

@reviewbybees

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

cliffmeyers commented:
I have one or two more small tweaks to make to this ticket but shouldn't have a major impact on overall code:
- [ ] Animations for cards that are added / removed
- [ ] Make the pipeline name (or possibly entire card) clickable for navigation

michaelneale commented on blueocean-dashboard/src/main/js/components/PipelinePage.jsx:
```diff
@@ -44,7 +44,11 @@ export default class PipelinePage extends Component {
                             <Link to={activityUrl}>{name}</Link>
                         </h1>
-                        <Favorite className="dark-yellow" />
+                        <Extensions.Renderer
```
👍🏻 moar extension points

kzantow commented:
:bee: after merge issue resolved (though, as we discussed earlier, the redux integration is going down a very ugly path)

michaelneale commented:
It looks good, but I wasn't able to click on the favourite for a branch in the details page. It instead selects the whole row and shows be the results for the latest run of that branch, so a bit of an 🐜

cliffmeyers commented:
@michaelneale this is fixed in 29c8770

cliffmeyers commented:
This should now be ready for final review and merge. Changes made since last review:

* Fixed bug where user could not unfavorite a multibranch pipeline from the dashboard card stack:  (see 0624b7d) https://issues.jenkins-ci.org/browse/JENKINS-36968
* Fixed bug where clicking on a favorite star on Branches tab was launching run details (see 29c8770)
* Linkify'd card name to open Run Details (see cbb7fc9)

I had to split off the animation work due to a blocking bug, work is tracked in these tickets:
https://issues.jenkins-ci.org/browse/JENKINS-37006
https://issues.jenkins-ci.org/browse/JENKINS-37007

kzantow commented on blueocean-dashboard/src/main/js/components/Branches.jsx:
```diff
@@ -3,6 +3,7 @@ import { CommitHash, ReadableDate } from '@jenkins-cd/design-language';
 import { LiveStatusIndicator, WeatherIcon } from '@jenkins-cd/design-language';
 import Extensions from '@jenkins-cd/js-extensions';
 import RunPipeline from './RunPipeline.jsx';
+import { StopPropagation } from './StopPropagation';
```
Hooray! :)

michaelneale replied:
very naaace.

kzantow commented on blueocean-personalization/src/main/js/components/FavoritePipelineHeader.jsx:
```diff
@@ -0,0 +1,18 @@
+
+    render() {
+        return (
+            <FavoritePipeline { ...this.props } className="dark-yellow" />
```
class name :(

kzantow commented:
🐝  still LGTM

michaelneale commented:
:bee: yes this works great. Thanks so much!
